pkg/controller/machinedeployment/sync.go

```diff
@@ -0,0 +1,445 @@
+		dutil.SetReplicasAnnotations(msCopy, *(deployment.Spec.Replicas), *(deployment.Spec.Replicas)+dutil.MaxSurge(*deployment))
+		ms, err = dc.machineClient.ClusterV1alpha1().MachineSets(msCopy.Namespace).Update(msCopy)
+		if err == nil && sizeNeedsUpdate {
+			scaled = true
+		}
+	}
+	return scaled, ms, err
+}
+
+// cleanupDeployment is responsible for cleaning up a deployment ie. retains all but the latest N old machine sets
+// where N=d.Spec.RevisionHistoryLimit. Old machine sets are older versions of the machinetemplate of a deployment kept
+// around by default 1) for historical reasons and 2) for the ability to rollback a deployment.
+func (dc *MachineDeploymentControllerImpl) cleanupDeployment(oldMSs []*v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment) error {
+	if deployment.Spec.RevisionHistoryLimit == nil {
+		return nil
+	}
+
+	// Avoid deleting machine set with deletion timestamp set
+	aliveFilter := func(ms *v1alpha1.MachineSet) bool {
+		return ms != nil && ms.ObjectMeta.DeletionTimestamp == nil
+	}
+	cleanableMSes := dutil.FilterMachineSets(oldMSs, aliveFilter)
+
+	diff := int32(len(cleanableMSes)) - *deployment.Spec.RevisionHistoryLimit
+	if diff <= 0 {
+		return nil
+	}
+
+	sort.Sort(dutil.MachineSetsByCreationTimestamp(cleanableMSes))
+	glog.V(4).Infof("Looking to cleanup old machine sets for deployment %q", deployment.Name)
+
+	for i := int32(0); i < diff; i++ {
+		ms := cleanableMSes[i]
+		// Avoid delete machine set with non-zero replica counts
+		if ms.Status.Replicas != 0 || *(ms.Spec.Replicas) != 0 || ms.Generation > ms.Status.ObservedGeneration || ms.DeletionTimestamp != nil {
+			continue
+		}
+		glog.V(4).Infof("Trying to cleanup machine set %q for deployment %q", ms.Name, deployment.Name)
+		if err := dc.machineClient.ClusterV1alpha1().MachineSets(ms.Namespace).Delete(ms.Name, nil); err != nil && !errors.IsNotFound(err) {
+			// Return error instead of aggregating and continuing DELETEs on the theory
+			// that we may be overloading the api server.
+			return err
+		}
+	}
+
+	return nil
+}
+
+// syncDeploymentStatus checks if the status is up-to-date and sync it if necessary
+func (dc *MachineDeploymentControllerImpl) syncDeploymentStatus(allMSs []*v1alpha1.MachineSet, newMS *v1alpha1.MachineSet, d *v1alpha1.MachineDeployment) error {
+	newStatus := calculateStatus(allMSs, newMS, d)
+
+	if reflect.DeepEqual(d.Status, newStatus) {
+		return nil
+	}
+
+	newDeployment := d
+	newDeployment.Status = newStatus
+	_, err := dc.machineClient.ClusterV1alpha1().MachineDeployments(newDeployment.Namespace).UpdateStatus(newDeployment)
+	return err
+}
+
+// calculateStatus calculates the latest status for the provided deployment by looking into the provided machine sets.
+func calculateStatus(allMSs []*v1alpha1.MachineSet, newMS *v1alpha1.MachineSet, deployment *v1alpha1.MachineDeployment) v1alpha1.MachineDeploymentStatus {
+	availableReplicas := dutil.GetAvailableReplicaCountForMachineSets(allMSs)
```

Comment:
PR mentioned MachineSet Status not implemented yet, is this still accurate and the code is just prepared for when AvailableReplicas is properly set?

Reply:
This is still accurate and you are correct.
The MachineSet Status has the fields for AvailableReplicas, but the MachineSet controller currently does not populate it.
The method pulls value out of the MachineSet Status, but since it isn't populated, it always returns 0, which is the default.